dali/pipeline/operators/geometric/bb_flip_test.cc

```diff
@@ -33,213 +38,151 @@ constexpr int kBbStructSize = 4;
 };
 
 
-/**
- * Injects either input values of test std::arrays (i.e. left-hand
- * Rois) or anticipated output values, which is the reference data.
- * @tparam DataIdx Index of tuple contained in TestData. Specifies
- *                 which values to inject
- */
-template<typename DataType, int DataIdx>
-void InjectTestData(const TestData &test_data, DataType *destination) {
-  for (const auto &it : test_data) {
-    auto _it = std::get<DataIdx>(it).data();
-    std::memcpy(destination, _it, kBbStructSize * sizeof(DataType));
-    destination += kBbStructSize;
+template<size_t N>
+const TestSample &FindSample(const TestSample (&dataset)[N], const Roi &roi) {
+  for (auto &sample : dataset) {
+    if (sample[0] == roi) {
+      return sample;
+    }
   }
+  DALI_FAIL("TestSample for provided `roi` has not been found");
 }
 
-}  // namespace
 
-template<typename ImageType>
-class BbFlipTest : public DALISingleOpTest<ImageType> {
- protected:
-  std::vector<TensorList<CPUBackend> *>
-  Reference(const std::vector<TensorList<CPUBackend> *> &inputs, DeviceWorkspace *ws) override {
-    using DataType = float;
-    auto batch = new TensorList<CPUBackend>();
-    batch->Resize(new_batch_size_);
-    auto *batch_data = batch->template mutable_data<DataType>();
-
-    DALI_ENFORCE(!(flip_type_horizontal_ && flip_type_vertical_), "No test data for combined case");
-
-    if (flip_type_vertical_) {
-      InjectTestData<DataType, 2>(*test_data_, batch_data);
-    } else if (flip_type_horizontal_) {
-      InjectTestData<DataType, 1>(*test_data_, batch_data);
-    } else {
-      InjectTestData<DataType, 0>(*test_data_, batch_data);
+/**
+ * Create TensorList from TestSample data
+ *
+ * @tparam Idx index of data @see TestSample
+ * @tparam Backend
+ * @tparam N size of TestSample data
+ * @param sample
+ * @return
+ */
+template<size_t Idx, typename Backend, size_t N>
+std::unique_ptr<TensorList<Backend>> ToTensorList(const TestSample (&sample)[N]) {
+  std::unique_ptr<TensorList<Backend>> tl(new TensorList<Backend>());
+  tl->Resize({N, {kBbStructSize}});
+
+  auto ptr = tl->template mutable_data<float>();
+  for (size_t j = 0; j < N; j++) {
+    for (size_t i = 0; i < kBbStructSize; i++) {
+      *ptr++ = sample[j][Idx][i];
     }
+  }
+  return tl;
+}
 
-    vector<TensorList<CPUBackend> *> ret(1);
-    ret[0] = batch;
 
-    return ret;
+template<typename Backend>
+Roi FromTensorWrapper(TensorListWrapper tw) {
+  auto tl = tw.get<Backend>();
+  assert(tl->size() >= kBbStructSize);
+  auto ptr = tl->template data<float>();
+  Roi roi;
+  for (float &val : roi) {
+    val = *ptr++;
   }
+  return roi;
+}
 
 
-  template<typename Backend>
-  void LoadBbData(TensorList<Backend> &batch, const TestData *input_data) noexcept {  // NOLINT
-    test_data_ = input_data;
+template<bool Ltrb>
+void Verify(TensorListWrapper input, TensorListWrapper output, Arguments args) {
+  auto input_roi = FromTensorWrapper<CPUBackend>(input);
```

Comment:
You only check the first element in the list.

Reply:
Done